### ISSUE TITLE
Investigate SchemaParser.parse_file raw_renderer.render_data performance.

### ISSUE THREAD

**Issue by gshank** [fishtown-analytics/dbt#3091]
This is code that looks for cycles in the DAG. It currently consumes 19% of the parsing time on our large test project. From the SchemaParse in core/dbt/parser/schemas.py, parse_file method calls raw_renderer.render_data, and that uses the 'deep_map' utility function, which is very slow.

Investigate and see if we can do this more efficiently.

**Comment by jtcohen6** [comment]
Uncertainties:
- Can we model the current weird structure of parsed YAML files in Rust? (Can the python implementation of deep_map be more opinionated as to the structure of the thing it's mapping?)
- Can we use a Rust Jinja lookalike to render strings instead of true python Jinja within deep_map, and return the same results?

This feels like a good candidate for Rust rework because:
- it's a really parallelizable task
- it's small in the sense of being self-contained within dbt's codebase.

A natural extension from here would be YAML file loading, which is a trivial task but touches many more files.

**Comment by nathaniel-may** [comment]
After some investigation I am closing this ticket because there are no easy wins in python here. If we decide to rework some of this code into rust we will use a new ticket for that.

Findings:
- _Can we model the current weird structure of parsed YAML files in Rust?_ No. The Python ABI needs more information than the type signature of `deep_map` provides.
- _Can we use a Rust Jinja lookalike to render strings instead of true python Jinja within deep_map, and return the same results?_ Yes, but this will be a much bigger and riskier change for users. When this is eventually attempted it will be tracked in a separate ticket.
- `deep_map` is running over every string in the project files to see if they are jinja. If a given string might be jinja based on a simple regex check, it runs the templating engine over the string. The slow parts of this function are the jinja templating itself, and the fact we just call it many times in serial.
- There are two opportunities for parallelization: 1. Each file could be `deep_map`ped in parallel. 2. Each file could traverse its tree-structure in parallel. 1 is a good candidate, 2 is not because these files are not expected to be particularly enormous.
